src/bids_validator/context.py

```diff
@@ -129,14 +130,41 @@ def dataset_description(self) -> Namespace:
     @cached_property
     def modalities(self) -> list[str]:
         """List of modalities found in the dataset."""
-        ...
-        return []
+        result = set()
+
+        modalities = self.schema.rules.modalities
+        for datatype in self.datatypes:
+            for mod_name, mod_dtypes in modalities.items():
+                if datatype in mod_dtypes.datatypes:
+                    result.add(mod_name)
+
+        return list(result)
 
     @cached_property
     def datatypes(self) -> list[str]:
         """List of datatypes found in the dataset."""
-        ...
-        return []
+        datatypes = self.schema.objects.datatypes
+        result = self.find_datatypes(self.tree, datatypes)
+
+        return list(result)
+
+    def find_datatypes(
+        self, tree: FileTree, datatypes: Namespace, result: set[str] | None = None
```

Comment:
I would move this method off the class into a standalone function. I don't think the functionality would ever be useful; if I want to find datatypes in a different `FileTree`, I would just use `Dataset(tree, schema).datatypes`.